File: src/game_learning_runtime/run_context.py
"""Verify invocation-scoped configuration selected by the standalone GLR CLI."""

from __future__ import annotations

import hashlib
import json
import os
import re
import sys
from collections.abc import Mapping
from dataclasses import dataclass
from pathlib import Path, PurePosixPath
from types import MappingProxyType
from typing import Any

from game_learning_runtime.project import GLRProject

if sys.version_info >= (3, 11):
    import tomllib
else:
    import tomli as tomllib  # type: ignore[import-not-found]

RUN_CONTEXT_SCHEMA_VERSION = "glr.run-context.v1"
_IDENTIFIER = re.compile(r"^[A-Za-z0-9][A-Za-z0-9_.-]{0,127}$")
_MAX_FILE_BYTES = 1024 * 1024
_MAX_ENV_BYTES = 24 * 1024


@dataclass(frozen=True, slots=True)
class RunContextFile:
    owner: str
    path: str
    schema_version: str
    sha256: str
    size_bytes: int

    def to_mapping(self) -> dict[str, object]:
        return {
            "owner": self.owner,
            "path": self.path,
            "schema_version": self.schema_version,
            "sha256": self.sha256,
            "size_bytes": self.size_bytes,
        }


@dataclass(frozen=True, slots=True)
class RunContextSource:
    path: str
    sha256: str
    size_bytes: int

    def to_mapping(self) -> dict[str, object]:
        return {
            "path": self.path,
            "sha256": self.sha256,
            "size_bytes": self.size_bytes,
        }


@dataclass(frozen=True, slots=True)
class RunContext:
    context_id: str
    environment_id: str
    protocol_version: str
    labels: Mapping[str, str]
    source: RunContextSource
    inputs: tuple[RunContextFile, ...]
    context_sha256: str
    schema_version: str = RUN_CONTEXT_SCHEMA_VERSION

    def unsigned_mapping(self) -> dict[str, object]:
        return {
            "schema_version": self.schema_version,
            "context_id": self.context_id,
            "environment_id": self.environment_id,
            "protocol_version": self.protocol_version,
            "labels": dict(sorted(self.labels.items())),
            "source": self.source.to_mapping(),
            "inputs": [item.to_mapping() for item in self.inputs],
        }

    def to_mapping(self) -> dict[str, object]:
        return {**self.unsigned_mapping(), "context_sha256": self.context_sha256}

    def to_json(self) -> str:
        return _compact_json(self.to_mapping())

    def verify(self, root: Path) -> None:
        _verify_identity(root, self.source.path, self.source.sha256, self.source.size_bytes)
        for item in self.inputs:
            _verify_identity(root, item.path, item.sha256, item.size_bytes)
        actual = _digest(_compact_json(self.unsigned_mapping()).encode())
        if actual != self.context_sha256:
            raise ValueError("run context identity changed in memory")


def load_run_context(
    project: GLRProject,
    path: str | Path | None = None,
    *,
    environment: Mapping[str, str] | None = None,
) -> RunContext:
    """Load a local context deliberately or verify the CLI-inherited receipt."""

    if environment is not None:
        if path is not None:
            raise ValueError("choose a local run context path or inherited environment")
        return _from_environment(project, environment)
    if path is None:
        raise ValueError("run context path is required")
    return _from_file(project, Path(path))


def load_inherited_run_context(
    project: GLRProject, environment: Mapping[str, str] | None = None
) -> RunContext | None:
    """Return no selection when both variables are absent; reject partial receipts."""

    source = os.environ if environment is None else environment
    present = [name in source for name in ("GLR_RUN_CONTEXT", "GLR_RUN_CONTEXT_SHA256")]
    if not any(present):
        return None
    if not all(present):
        raise ValueError("incomplete inherited run context environment")
    return _from_environment(project, source)


def _from_file(project: GLRProject, requested: Path) -> RunContext:
    source_path, source_relative = _project_file(project.root, requested)
    source_bytes = _read(source_path)
    try:
        raw = tomllib.loads(source_bytes.decode("utf-8"))
    except (UnicodeDecodeError, tomllib.TOMLDecodeError) as error:
        raise ValueError("run context must be valid UTF-8 TOML") from error
    value = _strict_mapping(
        raw,
        required={"schema_version", "context_id", "environment_id", "protocol_version", "inputs"},
        optional={"labels"},
        path="run context",
    )
    if value["schema_version"] != RUN_CONTEXT_SCHEMA_VERSION:
        raise ValueError(f"run context schema_version must be {RUN_CONTEXT_SCHEMA_VERSION!r}")
    context_id = _identifier(value["context_id"], "context_id")
    environment_id = _identifier(value["environment_id"], "environment_id")
    protocol_version = _text(value["protocol_version"], "protocol_version", 128)
    if environment_id != project.environment_id or protocol_version != project.protocol_version:
        raise ValueError("run context environment or protocol does not match the project")
    labels_value = value.get("labels", {})
    if not isinstance(labels_value, Mapping) or len(labels_value) > 32:
        raise ValueError("run context labels must be an object with at most 32 entries")
    labels = {
        _identifier(name, "label name"): _text(label, "label value", 256)
        for name, label in labels_value.items()
    }
    raw_inputs = value["inputs"]
    if not isinstance(raw_inputs, list) or not 1 <= len(raw_inputs) <= 64:
        raise ValueError("run context must contain 1-64 inputs")
    owners: set[str] = set()
    paths = {source_relative}
    inputs: list[RunContextFile] = []
    for index, raw_input in enumerate(raw_inputs):
        item = _strict_mapping(
            raw_input,
            required={"owner", "path", "schema_version"},
            optional=set(),
            path=f"run context inputs[{index}]",
        )
        owner = _identifier(item["owner"], "input owner")
        if owner in owners:
            raise ValueError(f"duplicate run context input owner: {owner}")
        owners.add(owner)
        expected_schema = _text(item["schema_version"], "input schema_version", 128)
        input_path, relative = _project_file(project.root, Path(item["path"]))
        if relative in paths:
            raise ValueError(f"duplicate run context input path: {relative}")
        paths.add(relative)
        data = _read(input_path)
        _verify_schema(input_path, data, expected_schema)
        inputs.append(RunContextFile(owner, relative, expected_schema, _digest(data), len(data)))
    source = RunContextSource(source_relative, _digest(source_bytes), len(source_bytes))
    frozen_inputs = tuple(inputs)
    frozen_labels = MappingProxyType(dict(sorted(labels.items())))
    provisional = RunContext(
        context_id=context_id,
        environment_id=environment_id,
        protocol_version=protocol_version,
        labels=frozen_labels,
        source=source,
        inputs=frozen_inputs,
        context_sha256="",
    )
    digest = _digest(_compact_json(provisional.unsigned_mapping()).encode())
    context = RunContext(
        context_id=context_id,
        environment_id=environment_id,
        protocol_version=protocol_version,
        labels=frozen_labels,
        source=source,
        inputs=frozen_inputs,
        context_sha256=digest,
    )
    if len(context.to_json().encode()) > _MAX_ENV_BYTES:
        raise ValueError("run context exceeds the 24 KiB role environment limit")
    return context


def _from_environment(project: GLRProject, environment: Mapping[str, str]) -> RunContext:
    try:
        raw_json = environment["GLR_RUN_CONTEXT"]
        expected_digest = environment["GLR_RUN_CONTEXT_SHA256"]
    except KeyError as error:
        raise ValueError("incomplete inherited run context environment") from error
    if len(raw_json.encode()) > _MAX_ENV_BYTES:
        raise ValueError("inherited run context exceeds 24 KiB")
    try:
        raw = json.loads(raw_json, object_pairs_hook=_unique_object)
    except json.JSONDecodeError as error:
        raise ValueError("inherited run context must be valid JSON") from error
    value = _strict_mapping(
        raw,
        required={
            "schema_version",
            "context_id",
            "environment_id",
            "protocol_version",
            "labels",
            "source",
            "inputs",
            "context_sha256",
        },
        optional=set(),
        path="inherited run context",
    )
    source = _strict_mapping(
        value["source"],
        required={"path", "sha256", "size_bytes"},
        optional=set(),
        path="run context source",
    )
    raw_inputs = value["inputs"]
    if not isinstance(raw_inputs, list):
        raise ValueError("run context inputs must be an array")
    inputs = tuple(_context_file(item, index) for index, item in enumerate(raw_inputs))
    if not 1 <= len(inputs) <= 64:
        raise ValueError("run context must contain 1-64 inputs")
    if len({item.owner for item in inputs}) != len(inputs):
        raise ValueError("duplicate inherited run context input owner")
    if len({item.path for item in inputs}) != len(inputs):
        raise ValueError("duplicate inherited run context input path")
    labels = value["labels"]
    if not isinstance(labels, Mapping) or any(
        not isinstance(name, str) or not isinstance(label, str) for name, label in labels.items()
    ):
        raise ValueError("run context labels must contain string keys and values")
    if len(labels) > 32:
        raise ValueError("run context labels exceed 32 entries")
    context = RunContext(
        schema_version=_text(value["schema_version"], "schema_version", 128),
        context_id=_identifier(value["context_id"], "context_id"),
        environment_id=_identifier(value["environment_id"], "environment_id"),
        protocol_version=_text(value["protocol_version"], "protocol_version", 128),
        labels=MappingProxyType(
            {
                _identifier(name, "label name"): _text(label, "label value", 256)
                for name, label in labels.items()
            }
        ),
        source=RunContextSource(
            path=_text(source["path"], "source path", 512),
            sha256=_sha256(source["sha256"], "source sha256"),
            size_bytes=_size(source["size_bytes"], "source size_bytes"),
        ),
        inputs=inputs,
        context_sha256=str(value["context_sha256"]),
    )
    if context.schema_version != RUN_CONTEXT_SCHEMA_VERSION:
        raise ValueError("unsupported inherited run context schema")
    if (
        context.environment_id != project.environment_id
        or context.protocol_version != project.protocol_version
    ):
        raise ValueError("inherited run context does not match the project")
    if context.context_sha256 != expected_digest:
        raise ValueError("inherited run context digest does not match its environment receipt")
    context.verify(project.root)
    return context


def _context_file(value: object, index: int) -> RunContextFile:
    item = _strict_mapping(
        value,
        required={"owner", "path", "schema_version", "sha256", "size_bytes"},
        optional=set(),
        path=f"run context inputs[{index}]",
    )
    return RunContextFile(
        owner=_identifier(item["owner"], "input owner"),
        path=_text(item["path"], "input path", 512),
        schema_version=_text(item["schema_version"], "input schema_version", 128),
        sha256=_sha256(item["sha256"], "input sha256"),
        size_bytes=_size(item["size_bytes"], "input size_bytes"),
    )


def _project_file(root: Path, requested: Path) -> tuple[Path, str]:
    portable = PurePosixPath(requested.as_posix())
    if requested.is_absolute() or any(part in {"", ".", ".."} for part in portable.parts):
        raise ValueError("run context paths must be portable project-relative paths")
    relative = portable.as_posix()
    if len(relative.encode()) > 512:
        raise ValueError("run context path exceeds 512 bytes")
    current = root
    for part in portable.parts:
        current = current / part
        if current.is_symlink():
            raise ValueError("run context paths cannot contain links")
    resolved = current.resolve()
    if not resolved.is_relative_to(root) or not resolved.is_file():
        raise FileNotFoundError(f"run context file is missing: {relative}")
    return resolved, relative


def _read(path: Path) -> bytes:
    value = path.read_bytes()
    if len(value) > _MAX_FILE_BYTES:
        raise ValueError(f"run context file exceeds 1 MiB: {path}")
    return value


def _verify_identity(root: Path, relative: str, expected: str, expected_size: int) -> None:
    path, normalized = _project_file(root, Path(relative))
    data = _read(path)
    if normalized != relative or len(data) != expected_size or _digest(data) != expected:
        raise ValueError(f"frozen run context input changed: {relative}")


def _verify_schema(path: Path, data: bytes, expected: str) -> None:
    try:
        if path.suffix == ".json":
            value = json.loads(data)
        elif path.suffix == ".toml":
            value = tomllib.loads(data.decode("utf-8"))
        else:
            raise ValueError("run context inputs must be JSON or TOML")
    except (UnicodeDecodeError, json.JSONDecodeError, tomllib.TOMLDecodeError) as error:
        raise ValueError(f"invalid run context input: {path}") from error
    if not isinstance(value, Mapping) or value.get("schema_version") != expected:
        raise ValueError(f"run context input schema_version must be {expected!r}")


def _strict_mapping(
    value: object, *, required: set[str], optional: set[str], path: str
) -> Mapping[str, Any]:
    if not isinstance(value, Mapping) or any(not isinstance(key, str) for key in value):
        raise TypeError(f"{path} must be an object with string keys")
    missing = sorted(required - set(value))
    unexpected = sorted(set(value) - required - optional)
    if missing or unexpected:
        raise ValueError(f"{path} has missing={missing} unexpected={unexpected} fields")
    return value


def _unique_object(items: list[tuple[str, object]]) -> dict[str, object]:
    result: dict[str, object] = {}
    for key, value in items:
        if key in result:
            raise ValueError(f"duplicate JSON key in inherited run context: {key}")
        result[key] = value
    return result


def _identifier(value: object, path: str) -> str:
    if not isinstance(value, str) or _IDENTIFIER.fullmatch(value) is None:
        raise ValueError(f"{path} must be a portable identifier")
    return value


def _text(value: object, path: str, maximum: int) -> str:
    if (
        not isinstance(value, str)
        or not value
        or len(value.encode()) > maximum
        or any(ord(character) < 32 for character in value)
    ):
        raise ValueError(f"{path} must be bounded printable text")
    return value


def _sha256(value: object, path: str) -> str:
    if not isinstance(value, str) or re.fullmatch(r"[0-9a-f]{64}", value) is None:
        raise ValueError(f"{path} must be a lowercase SHA-256 digest")
    return value


def _size(value: object, path: str) -> int:
    if not isinstance(value, int) or isinstance(value, bool) or not 0 <= value <= _MAX_FILE_BYTES:
        raise ValueError(f"{path} must be a bounded byte size")
    return value


def _compact_json(value: object) -> str:
    return json.dumps(value, ensure_ascii=False, separators=(",", ":"))


def _digest(value: bytes) -> str:
    return hashlib.sha256(value).hexdigest()


__all__ = [
    "RUN_CONTEXT_SCHEMA_VERSION",
    "RunContext",
    "RunContextFile",
    "RunContextSource",
    "load_inherited_run_context",
    "load_run_context",
]

File: crates/glr-cli/src/run_context.rs
use std::collections::{BTreeMap, HashSet};
use std::fs;
use std::path::{Component, Path, PathBuf};

use serde::{Deserialize, Serialize};
use serde_json::Value;
use sha2::{Digest, Sha256};

use crate::error::{Error, Result};
use crate::project::Project;

pub const SCHEMA_VERSION: &str = "glr.run-context.v1";
const MAX_CONTEXT_BYTES: usize = 1024 * 1024;
const MAX_CONTEXT_ENV_BYTES: usize = 24 * 1024;
const MAX_INPUTS: usize = 64;
const MAX_LABELS: usize = 32;

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
struct ContextFile {
    schema_version: String,
    context_id: String,
    environment_id: String,
    protocol_version: String,
    #[serde(default)]
    labels: BTreeMap<String, String>,
    inputs: Vec<InputReference>,
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
struct InputReference {
    owner: String,
    path: String,
    schema_version: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct FileIdentity {
    owner: String,
    path: String,
    schema_version: String,
    sha256: String,
    size_bytes: usize,
}

#[derive(Debug, Clone, Serialize)]
struct ContextIdentity {
    path: String,
    sha256: String,
    size_bytes: usize,
}

#[derive(Debug, Clone, Serialize)]
struct UnsignedContext {
    schema_version: &'static str,
    context_id: String,
    environment_id: String,
    protocol_version: String,
    labels: BTreeMap<String, String>,
    source: ContextIdentity,
    inputs: Vec<FileIdentity>,
}

#[derive(Debug, Clone)]
pub struct RunContext {
    unsigned: UnsignedContext,
    context_sha256: String,
}

impl RunContext {
    pub fn load(project: &Project, requested: &Path) -> Result<Self> {
        let (source_path, source_relative) = regular_project_file(&project.root, requested)?;
        let source_bytes = read_bounded(&source_path)?;
        let text = std::str::from_utf8(&source_bytes)
            .map_err(|_| invalid("run context must be UTF-8 TOML"))?;
        let value: ContextFile = toml::from_str(text)
            .map_err(|error| invalid(format!("invalid run context TOML: {error}")))?;
        if value.schema_version != SCHEMA_VERSION {
            return Err(invalid(format!(
                "run context schema_version must be {SCHEMA_VERSION:?}"
            )));
        }
        identifier(&value.context_id, "context_id")?;
        if value.environment_id != project.environment_id {
            return Err(invalid(
                "run context environment_id does not match the project",
            ));
        }
        if value.protocol_version != project.protocol_version {
            return Err(invalid(
                "run context protocol_version does not match the project",
            ));
        }
        if value.inputs.is_empty() || value.inputs.len() > MAX_INPUTS {
            return Err(invalid(format!(
                "run context must contain 1-{MAX_INPUTS} inputs"
            )));
        }
        if value.labels.len() > MAX_LABELS {
            return Err(invalid(format!(
                "run context cannot contain more than {MAX_LABELS} labels"
            )));
        }
        for (name, label) in &value.labels {
            identifier(name, "run context label name")?;
            printable(label, "run context label value", 256)?;
        }
        let mut owners = HashSet::new();
        let mut paths = HashSet::new();
        let mut inputs = Vec::with_capacity(value.inputs.len());
        for reference in value.inputs {
            identifier(&reference.owner, "run context input owner")?;
            printable(
                &reference.schema_version,
                "run context input schema_version",
                128,
            )?;
            if !owners.insert(reference.owner.clone()) {
                return Err(invalid(format!(
                    "duplicate run context input owner: {}",
                    reference.owner
                )));
            }
            let requested = Path::new(&reference.path);
            let (path, relative) = regular_project_file(&project.root, requested)?;
            if relative == source_relative || !paths.insert(relative.clone()) {
                return Err(invalid(format!(
                    "duplicate run context input path: {relative}"
                )));
            }
            let bytes = read_bounded(&path)?;
            verify_schema(&path, &bytes, &reference.schema_version)?;
            inputs.push(FileIdentity {
                owner: reference.owner,
                path: relative,
                schema_version: reference.schema_version,
                sha256: digest(&bytes),
                size_bytes: bytes.len(),
            });
        }
        let unsigned = UnsignedContext {
            schema_version: SCHEMA_VERSION,
            context_id: value.context_id,
            environment_id: value.environment_id,
            protocol_version: value.protocol_version,
            labels: value.labels,
            source: ContextIdentity {
                path: source_relative,
                sha256: digest(&source_bytes),
                size_bytes: source_bytes.len(),
            },
            inputs,
        };
        let context_sha256 = digest(&serde_json::to_vec(&unsigned)?);
        let context = Self {
            unsigned,
            context_sha256,
        };
        if context.json()?.len() > MAX_CONTEXT_ENV_BYTES {
            return Err(invalid(
                "run context exceeds the 24 KiB role environment limit",
            ));
        }
        Ok(context)
    }

    pub fn value(&self) -> Result<Value> {
        let mut value = serde_json::to_value(&self.unsigned)?;
        value
            .as_object_mut()
            .expect("run context serializes as an object")
            .insert(
                "context_sha256".into(),
                Value::String(self.context_sha256.clone()),
            );
        Ok(value)
    }

    pub fn json(&self) -> Result<String> {
        Ok(serde_json::to_string(&self.value()?)?)
    }

    pub fn digest(&self) -> &str {
        &self.context_sha256
    }

    pub fn verify(&self, root: &Path) -> Result<()> {
        verify_identity(
            root,
            &self.unsigned.source.path,
            &self.unsigned.source.sha256,
            self.unsigned.source.size_bytes,
        )?;
        for input in &self.unsigned.inputs {
            verify_identity(root, &input.path, &input.sha256, input.size_bytes)?;
        }
        let expected = digest(&serde_json::to_vec(&self.unsigned)?);
        if expected != self.context_sha256 {
            return Err(Error::Contract(
                "run context identity changed in memory".into(),
            ));
        }
        Ok(())
    }
}

fn verify_identity(
    root: &Path,
    relative: &str,
    expected: &str,
    expected_size: usize,
) -> Result<()> {
    let (path, normalized) = regular_project_file(root, Path::new(relative))?;
    if normalized != relative {
        return Err(Error::Contract("run context path identity changed".into()));
    }
    let bytes = read_bounded(&path)?;
    if bytes.len() != expected_size || digest(&bytes) != expected {
        return Err(Error::Contract(format!(
            "frozen run context input changed before role execution: {relative}"
        )));
    }
    Ok(())
}

fn verify_schema(path: &Path, bytes: &[u8], expected: &str) -> Result<()> {
    let actual = match path.extension().and_then(|value| value.to_str()) {
        Some("json") => serde_json::from_slice::<Value>(bytes)?
            .get("schema_version")
            .and_then(Value::as_str)
            .map(str::to_owned),
        Some("toml") => {
            let text = std::str::from_utf8(bytes)
                .map_err(|_| invalid("run context TOML input must be UTF-8"))?;
            toml::from_str::<toml::Value>(text)
                .map_err(|error| invalid(format!("invalid TOML context input: {error}")))?
                .get("schema_version")
                .and_then(toml::Value::as_str)
                .map(str::to_owned)
        }
        _ => return Err(invalid("run context inputs must be JSON or TOML")),
    };
    if actual.as_deref() != Some(expected) {
        return Err(invalid(format!(
            "run context input {} schema_version must be {expected:?}",
            path.display()
        )));
    }
    Ok(())
}

fn regular_project_file(root: &Path, requested: &Path) -> Result<(PathBuf, String)> {
    if requested.as_os_str().is_empty()
        || requested.is_absolute()
        || requested.components().any(|component| {
            matches!(
                component,
                Component::CurDir
                    | Component::ParentDir
                    | Component::Prefix(_)
                    | Component::RootDir
            )
        })
    {
        return Err(invalid("run context paths must be project-relative"));
    }
    let relative = requested.to_string_lossy().replace('\\', "/");
    if relative.len() > 512 {
        return Err(invalid("run context path exceeds 512 bytes"));
    }
    let mut current = root.to_path_buf();
    for component in requested.components() {
        if let Component::Normal(value) = component {
            current.push(value);
            if current.exists() && linked(&current)? {
                return Err(invalid(
                    "run context paths cannot contain links or reparse points",
                ));
            }
        }
    }
    if !current.is_file() || linked(&current)? {
        return Err(Error::Missing(current));
    }
    Ok((current, relative))
}

fn linked(path: &Path) -> Result<bool> {
    let metadata = fs::symlink_metadata(path)?;
    if metadata.file_type().is_symlink() {
        return Ok(true);
    }
    #[cfg(windows)]
    {
        use std::os::windows::fs::MetadataExt;
        const FILE_ATTRIBUTE_REPARSE_POINT: u32 = 0x400;
        if metadata.file_attributes() & FILE_ATTRIBUTE_REPARSE_POINT != 0 {
            return Ok(true);
        }
    }
    Ok(false)
}

fn read_bounded(path: &Path) -> Result<Vec<u8>> {
    let bytes = fs::read(path)?;
    if bytes.len() > MAX_CONTEXT_BYTES {
        return Err(invalid(format!(
            "run context file exceeds the 1 MiB limit: {}",
            path.display()
        )));
    }
    Ok(bytes)
}

fn identifier(value: &str, label: &str) -> Result<()> {
    if value.is_empty()
        || value.len() > 128
        || !value.as_bytes()[0].is_ascii_lowercase()
        || !value.bytes().all(|byte| {
            byte.is_ascii_lowercase() || byte.is_ascii_digit() || matches!(byte, b'-' | b'_' | b'.')
        })
    {
        return Err(invalid(format!(
            "{label} must start with a lowercase letter and contain lowercase ASCII letters, digits, '.', '_' or '-'"
        )));
    }
    Ok(())
}

fn printable(value: &str, label: &str, maximum: usize) -> Result<()> {
    if value.is_empty() || value.len() > maximum || value.chars().any(char::is_control) {
        return Err(invalid(format!(
            "{label} must be a bounded printable string"
        )));
    }
    Ok(())
}

fn digest(bytes: &[u8]) -> String {
    format!("{:x}", Sha256::digest(bytes))
}

fn invalid(message: impl Into<String>) -> Error {
    Error::Invalid(message.into())
}

pub fn metadata(project: &Project) -> Result<Value> {
    project
        .run_context
        .as_ref()
        .map(RunContext::value)
        .transpose()
        .map(|value| value.unwrap_or(Value::Null))
}

pub fn persist(
    project: &Project,
    store: &crate::store::Store,
    run_id: &str,
    run_dir: &Path,
) -> Result<()> {
    let Some(context) = &project.run_context else {
        return Ok(());
    };
    context.verify(&project.root)?;
    let path = run_dir.join("run-context.json");
    crate::contracts::write_json(&path, &context.value()?)?;
    store.register_artifact(
        run_id,
        "run-context.json",
        &path,
        "run-context",
        "application/json",
    )?;
    store.append_event(run_id, "context.selected", context.value()?)?;
    Ok(())
}
